### PR TITLE
np.float64("np.inf").astype(np.int32) is negative on x86 but positive on ppc64le

### PR DESCRIPTION
**Pull request by sandipmgiri**
- np.float64("np.inf").astype(np.int32) is negative on x86 but positive on ppc64le, that is the reason we 
  added this special case for ppc64le

- Discussed with numpy community and they mentioned as follows -
1) Doing this cast incurs undefined behaviour, so it's hard to say which platform is incorrect here.
    I'd argue that we should just issue a warning and leave the platform-specific behaviour.
2) This is also true of any C program that tries to cast float to int. If the author of a python package 
    included a C extension that does this cast, they still have the problem. 
   The same argument you use for "we should fix this is numpy, not in the user package" can be 
    applied to "we should fix this in C, not in numpy". Presumably for speed reasons, that argument was 
    rejected during the  design of C.
    Having said that, we already diverge from C behaviour in some place (eg, integer promotion for 
    arithmetic and comparison), so it wouldn't be unreasonable  to add special casing here.
    
Numpy link to relevant discussion - https://github.com/numpy/numpy/issues/9040
Also discussed this on Tensorflow , link to relevant discussion - https://github.com/tensorflow/tensorflow/issues/9360

### CODE REVIEW

**Comment by sandipmgiri** [comment]
Hi  @martinwicke, as discussed on thread https://github.com/tensorflow/tensorflow/issues/9360, I have created this PR.

**Comment by sandipmgiri** [comment]
For second (https://github.com/tensorflow/tensorflow/pull/10522/commits/c63cebb80a26cb7eeaf0afe095886488f1a8ac8a) commit details are as follows :

This is regarding failure of test case //tensorflow/core:platform_profile_utils_cpu_utils_test in tensorflow/core/platform/profile_utils/cpu_utils_test.cc file.
On ppc64le this test was failing for following two checks i.e. CHECK_GT(cpu_frequency, 0) and CHECK_NE(cpu_frequency, CpuUtils::INVALID_FREQUENCY) :

https://github.com/tensorflow/tensorflow/blob/v1.0.1/tensorflow/core/platform/profile_utils/cpu_utils_test.cc#L56-L57

```
TEST_F(CpuUtilsTest, CheckCycleCounterFrequency) {
  const int64 cpu_frequency = CpuUtils::GetCycleCounterFrequency();
  CHECK_GT(cpu_frequency, 0);
  CHECK_NE(cpu_frequency, CpuUtils::INVALID_FREQUENCY);
  if (DBG) {
    LOG(INFO) << "Cpu frequency = " << cpu_frequency;
  }
}
```

After invetigation found that - "cpu_frequency" value on x86 is "usigned" by default whereas on ppc64le it is "signed" by default, that is why this test was failing on ppc64le, PFA for log details 
[platform_profile_utils_cpu_utils_test failure_log.txt](https://github.com/tensorflow/tensorflow/files/1071075/platform_profile_utils_cpu_utils_test.failure_log.txt)


To make the results compatible with x86, some code changes were required on ppc64le. 
After this code changes, I ran this test and it is passing on ppc64le as well as on X86.

Please review this changes as well?
Thanks!

**Comment by martinwicke** [comment]
@prb12 friendly nudge?

File: tensorflow/core/platform/profile_utils/cpu_utils.cc [hunk 1]
```diff
@@ -53,7 +53,7 @@ static ICpuUtilsHelper* cpu_utils_helper_instance_ = nullptr;
                                        GetCycleCounterFrequency());
 }
 
-/* static */ int64 CpuUtils::GetCycleCounterFrequencyImpl() {
+/* static */ uint64 CpuUtils::GetCycleCounterFrequencyImpl() {
```
Comment (prb12) [inline tensorflow/core/platform/profile_utils/cpu_utils.cc]: What is the reason for this change?

Several of the return paths for the function return negative numbers (INVALID_FREQUENCY), some now with implicit casts to unsigned.  

The only use in the rest of this patch adds the opposide cast of INVALID_FREQUENCY back to unsigned.  

What am I missing here?

File: tensorflow/python/kernel_tests/cast_op_test.py [hunk 2]
```diff
@@ -146,9 +147,16 @@ def testInfNan(self):
     if sys.byteorder == "big":  
       self._compare(np.inf, np.int32, i4.max, False)  
       self._compare(np.inf, np.int64, i8.max, False)  
-    else:  
-      self._compare(np.inf, np.int32, i4.min, False)  
-      self._compare(np.inf, np.int64, i8.min, False)  
+    else:
+      #np.float64("np.inf").astype(np.int32) is negative on x86 but positive on ppc64le
```
Comment (prb12) [inline tensorflow/python/kernel_tests/cast_op_test.py]: Nit: Comments should have a space after the #

File: tensorflow/python/kernel_tests/cast_op_test.py [hunk 3]
```diff
@@ -146,9 +147,16 @@ def testInfNan(self):
...
+      #Numpy link to relevant discussion - https://github.com/numpy/numpy/issues/9040
+      #Tensorflow link to relevant discussion - https://github.com/tensorflow/tensorflow/issues/9360
+      if platform.machine() == "ppc64le":
+        self._compare(-np.inf, np.int32, i4.min, False)
```
Comment (prb12) [inline tensorflow/python/kernel_tests/cast_op_test.py]: As I read the posts linked to above, -np.inf is technically undefined?  
I think this is a little ugly, but fine as a temporary fix, and it doesn't affect other architectures.

**Review by prb12 — COMMENTED** [review]
Could you please provide some justification for the change to GetCycleCounterFrequency?

**Comment by sandipmgiri** [comment]
**Details for the change to GetCycleCounterFrequency :** 

This is regarding test case` //tensorflow/core:platform_profile_utils_cpu_utils_test`. While running this test on `ppc64le` platform `GetCycleCounterFrequency` function returning  unexpected  value (i.e. `-1` vs expected `greater than 1`)  ,which is ` "int64 cpu_frequency"` value
As per analysis here we found that  :  "`int64`"  value on `x86` is "unsigned" by default whereas on `ppc64le` it is "signed" by default. Because of that we were getting incorrect value on power.

To make the result compatible with `x86`, I have done this changes on `ppc64le`.
After code changes, I ran this test and now it is passing on` ppc64le` as well as on `X86`.

In first commit (https://github.com/tensorflow/tensorflow/commit/c63cebb80a26cb7eeaf0afe095886488f1a8ac8a) I had made generic changes . After that I realized that this test is passing successfully on` X86` without any change. Hence, now I have made power specific changes only in latest commit (i.e.  https://github.com/tensorflow/tensorflow/pull/10522/commits/c42d2f52a8237e7e73e29a5658bfe0c83baee414)

Please provide your comments on this and  let me know if you need any other details. Thanks!

File: tensorflow/core/platform/profile_utils/cpu_utils.cc (hunk 1 above)
Comment (drpngx) [inline tensorflow/core/platform/profile_utils/cpu_utils.cc]: @sandipmgiri could you address this?

File: tensorflow/python/kernel_tests/cast_op_test.py (hunk 3 above)
Comment (drpngx) [inline tensorflow/python/kernel_tests/cast_op_test.py]: @sandipmgiri could you fix the comment style?

**Comment by sandipmgiri** [comment]
@drpngx fixed comment style and made required changes. Please review and let me know if you need any additional details. Thanks!

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Thanks!

@prb12 good to go?

**Review by prb12 — APPROVED** [review]

**Comment by eric-wieser** [comment]
> "int64" value on x86 is "unsigned" by default

What exactly do you mean by this? `int64_t` is defined as a _"signed integer type with width of exactly 8, 16, 32 and 64 bits respectively"_. So I still have no understanding of why that change to `GetCycleCounterFrequency` is necessary

File: tensorflow/core/platform/profile_utils/cpu_utils_test.cc [hunk 4]
```diff
@@ -53,9 +53,15 @@ TEST_F(CpuUtilsTest, CheckGetCurrentClockCycle) {
 }
 
 TEST_F(CpuUtilsTest, CheckCycleCounterFrequency) {
-  const int64 cpu_frequency = CpuUtils::GetCycleCounterFrequency();
-  CHECK_GT(cpu_frequency, 0);
-  CHECK_NE(cpu_frequency, CpuUtils::INVALID_FREQUENCY);
+  #if defined(__powerpc__) || defined(__ppc__) && ( __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__)
+     const uint64 cpu_frequency = CpuUtils::GetCycleCounterFrequency();
+     CHECK_GT(cpu_frequency, 0);
+     CHECK_NE(cpu_frequency, unsigned(CpuUtils::INVALID_FREQUENCY));
```
Comment (eric-wieser) [inline tensorflow/core/platform/profile_utils/cpu_utils_test.cc]: Why are you casting to `unsigned`, which means `unsigned int`? Shouldn't you be casting to `int64`?
